### PR TITLE
feat: First drop of storage APIs for DataService and Wires

### PR DESCRIPTION
**Pull request by nicolatimeus**
Signed-off-by: Nicola Timeus <nicola.timeus@eurotech.com>

> **Note**: We are using the Conventional Commits convention for our pull request titles. Please take a look at the [PR title format document](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#submitting-the-changes) for the supported [types](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#type) and [scopes](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#scope).

Brief description of the PR. [e.g. Added `null` check on `object` to avoid `NullPointerException`]

Adds new APIs that allows to plug custom storage implementation to DataService and Wires.

* Added new APIs
* Adapted DataService to use the new APIs (it should still be possible to bind the DataService to an H2DbService that does not implement the new APIs)
* Adapted H2 Db service to implement the new APIs
* Added new Wire Record Store and Wire Record Query wire components that use the new APIs (the existing  database related wire components are kept as deprecated to keep backwards compatibility).

### CODE REVIEW

**Comment by nicolatimeus** [comment]
Should be ready for review now. The build is failing due to sonar reporting code duplication, it is not possible to reduce code duplication without dropping classes exported by internal packages causing breaking change.

File: kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/store/provider/MessageStore.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+ * See {@link StoredMessage} for a description of the stored message fields.
+ * 
+ * @since 2.5
+ * @noimplement This interface is not intended to be implemented by clients.
+ */
+@ProviderType
+public interface MessageStore {
+
+    /**
+     * Inserts a new message in the store. The implementation must se the value of
```
Comment (marcellorinaldo) [inline kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/message/store/provider/MessageStore.java]: The implementation must se the -> The implementation must se**t** the

**Review by marcellorinaldo — APPROVED** [review]

**Review by salvatore-coppola — APPROVED** [review]

**Review by marcellorinaldo — APPROVED** [review]

**Comment by MMaiero** [comment]
@pierantoniomerlino @mattdibi Any comments from you?

**Comment by pierantoniomerlino** [comment]
@MMaiero I'm checking the PR now. I had some discussion with Nicola about some topics.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/db/H2DbMessageStoreImpl.java [hunk 2]
```diff
@@ -1,52 +1,30 @@
...
- * available under the terms of the Eclipse Public License 2.0
- * which is available at https://www.eclipse.org/legal/epl-2.0/
- * 
- * SPDX-License-Identifier: EPL-2.0
- * 
- * Contributors:
- *  Eurotech
- *******************************************************************************/
-package org.eclipse.kura.core.data.store;
+package org.eclipse.kura.core.db;
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/db/H2DbMessageStoreImpl.java:13]: Copyright missing?

**Comment by pierantoniomerlino** [comment]
For me the PR is fine. As discussed with Nicola, I don't like the asymmetry between `QueryableWireRecordStoreProvider` and `WireRecordStoreProvider` and `WireRecordStore`. I'll search for an alternative...

**Review by pierantoniomerlino — APPROVED** [review]

**Comment by MMaiero** [comment]
> For me the PR is fine. As discussed with Nicola, I don't like the asymmetry between `QueryableWireRecordStoreProvider` and `WireRecordStoreProvider` and `WireRecordStore`. I'll search for an alternative...

So good to merge or better to wait?

**Comment by pierantoniomerlino** [comment]
@MMaiero I haven't a better idea about the names, so let's keep them as they are. Only the missing copyright should be added.

**Review by pierantoniomerlino — APPROVED** [review]

**Review by marcellorinaldo — APPROVED** [review]

**Pull request merged**